File: algoritmos_e_estruturas/sort/quick_sort.py



def quicksort_v1(array):
    if len(array) < 2:
        return array
    else:
        pivo = array[0]
        menores = [i for i in array[1:] if i <= pivo]
        maiores = [i for i in array[1:] if i > pivo]
        return quicksort_v1(menores) + [pivo] + quicksort_v1(maiores)
    

def quicksort(array, inicio, fim):
    if inicio < fim:
        # pegar o pivo, pegar o pivo e separar o array em left e right
        pivo = partition(array, inicio, fim)
        quicksort(array, pivo +1, fim) # lado direito - right
        quicksort(array, inicio, pivo - 1) # lado esquerdo - left

def partition(array, inicio, fim):
    pivo = array[fim] # ultimo elemento
    i = inicio -1 # index - comeca apontando para nada
    
    for j in range(inicio, fim):
        if array[j] <= pivo:
            i = i + 1 # Avançar barra amarela
            array[j], array[i] = array[i],array[j] 

    # colocar o pivo no lugar correto
    array[i+1], array[fim] = array[fim], array[i+1]

    return i +1 ## posição do pivo


inicio = 0
lista = [2,1,10,5,7]

print(lista)
tamanho = len(lista)
quicksort(lista, 0, tamanho-1)
print(lista)

File: algoritmos_e_estruturas/sort/quick_sort.cpp
#include <iostream>


using namespace std;
int partition(int vetor[], int left, int right){
   int pivo = vetor[right]; // Pivo ultimo elemento
   int i = left -1; // Index - apontando para o vazio
   int temp; // variavel temporaria p/ troca

   for (int j=left;j<right;j++){
        if (vetor[j] <= pivo){
            i++;
            temp = vetor[j];
            vetor[j] = vetor[i];
            vetor[i] = temp;
         }

   }
   // Colocor o pivo no lugar correto
   temp = vetor[i+1];
   vetor[i+1] = vetor[right];
   vetor[right] = temp;

   return i +1;
}

void quicksort(int vetor[], int left, int right) {
    int pivo;
    if (left < right)
    {
        pivo = partition(vetor, left, right);
        quicksort(vetor, left, pivo-1);
        quicksort(vetor,pivo+1,right);
    }
  
}

int main() {
    int vetor[] = {10, 15, 1, 5, 20, 7}; // um array de 6 elementos
    int tam = sizeof(vetor) / sizeof(vetor[0]);

    cout << "Array Desordenado: " << endl << "[ ";
    for (int i = 0; i < tam; i++) {
        cout << vetor[i] << " ";
    }
    cout << "]" << endl;


    quicksort(vetor, 0, tam - 1);

    cout << "Array ordenado: " << endl << "[ ";
    for (int i = 0; i < tam; i++) {
        cout << vetor[i] << " ";
    }
    cout << "]" << endl;

    return 0;
}
